# Data Visualization with Python

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

In [52]:
# Libraries to help with reading and manipulating data
import numpy as np
import pandas as pd

# Libraries to help with data visualization
import matplotlib.pyplot as plt
import seaborn as sns

# Command to tell Python to actually display the graphs
%matplotlib inline

In [22]:
df = pd.read_csv('data/Automobile.csv')
# df = pd.read_csv('/location on your computer/Automobile.csv')

In [23]:
df.head()

symboling  normalized_losses         make fuel_type aspiration  \
0          3                168  alfa-romero       gas        std   
1          3                168  alfa-romero       gas        std   
2          1                168  alfa-romero       gas        std   
3          2                164         audi       gas        std   
4          2                164         audi       gas        std   

  number_of_doors   body_style drive_wheels engine_location  wheel_base  ...  \
0             two  convertible          rwd           front        88.6  ...   
1             two  convertible          rwd           front        88.6  ...   
2             two    hatchback          rwd           front        94.5  ...   
3            four        sedan          fwd           front        99.8  ...   
4            four        sedan          4wd           front        99.4  ...   

   engine_size  fuel_system  bore  stroke compression_ratio horsepower  \
0          130         mpfi  3.47    2.68               9.0        111   
1          130         mpfi  3.47    2.68               9.0        111   
2          152         mpfi  2.68    3.47               9.0        154   
3          109         mpfi  3.19    3.40              10.0        102   
4          136         mpfi  3.19    3.40               8.0        115   

   peak_rpm city_mpg  highway_mpg  price  
0      5000       21           27  13495  
1      5000       21           27  16500  
2      5000       19           26  16500  
3      5500       24           30  13950  
4      5500       18           22  17450  

[5 rows x 26 columns]

In [24]:
df.shape

(201, 26)

- The data has 201 rows and 26 columns.

In [25]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 201 entries, 0 to 200
Data columns (total 26 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   symboling            201 non-null    int64  
 1   normalized_losses    201 non-null    int64  
 2   make                 201 non-null    object 
 3   fuel_type            201 non-null    object 
 4   aspiration           201 non-null    object 
 5   number_of_doors      201 non-null    object 
 6   body_style           201 non-null    object 
 7   drive_wheels         201 non-null    object 
 8   engine_location      201 non-null    object 
 9   wheel_base           201 non-null    float64
 10  length               201 non-null    float64
 11  width                201 non-null    float64
 12  height               201 non-null    float64
 13  curb_weight          201 non-null    int64  
 14  engine_type          201 non-null    object 
 15  number_of_cylinders  201 non-null    obj

- There are attributes of different types (*int*, *float*, *object*) in the data.

In [26]:
df.describe().T

count          mean          std      min      25%  \
symboling          201.0      0.840796     1.254802    -2.00     0.00   
normalized_losses  201.0    125.189055    33.572966    65.00   101.00   
wheel_base         201.0     98.797015     6.066366    86.60    94.50   
length             201.0    174.200995    12.322175   141.10   166.80   
width              201.0     65.889055     2.101471    60.30    64.10   
height             201.0     53.766667     2.447822    47.80    52.00   
curb_weight        201.0   2555.666667   517.296727  1488.00  2169.00   
engine_size        201.0    126.875622    41.546834    61.00    98.00   
bore               201.0      3.329701     0.268166     2.54     3.15   
stroke             201.0      3.261741     0.317875     2.07     3.11   
compression_ratio  201.0     10.164279     4.004965     7.00     8.60   
horsepower         201.0    103.263682    37.389372    48.00    70.00   
peak_rpm           201.0   5121.393035   479.624905  4150.00  4800.00   
city_mpg           201.0     25.179104     6.423220    13.00    19.00   
highway_mpg        201.0     30.686567     6.815150    16.00    25.00   
price              201.0  13207.129353  7947.066342  5118.00  7775.00   

                        50%       75%       max  
symboling              1.00      2.00      3.00  
normalized_losses    122.00    150.00    256.00  
wheel_base            97.00    102.40    120.90  
length               173.20    183.50    208.10  
width                 65.50     66.60     72.00  
height                54.10     55.50     59.80  
curb_weight         2414.00   2926.00   4066.00  
engine_size          120.00    141.00    326.00  
bore                   3.31      3.58      3.94  
stroke                 3.29      3.46      4.17  
compression_ratio      9.00      9.40     23.00  
horsepower            95.00    116.00    262.00  
peak_rpm            5200.00   5500.00   6600.00  
city_mpg              24.00     30.00     49.00  
highway_mpg           30.00     34.00     54.00  
price              10295.00  16500.00  45400.00

In [27]:
df.describe(include='all').T

count unique     top freq          mean          std  \
symboling            201.0    NaN     NaN  NaN      0.840796     1.254802   
normalized_losses    201.0    NaN     NaN  NaN    125.189055    33.572966   
make                   201     22  toyota   32           NaN          NaN   
fuel_type              201      2     gas  181           NaN          NaN   
aspiration             201      2     std  165           NaN          NaN   
number_of_doors        201      2    four  114           NaN          NaN   
body_style             201      5   sedan   94           NaN          NaN   
drive_wheels           201      3     fwd  118           NaN          NaN   
engine_location        201      2   front  198           NaN          NaN   
wheel_base           201.0    NaN     NaN  NaN     98.797015     6.066366   
length               201.0    NaN     NaN  NaN    174.200995    12.322175   
width                201.0    NaN     NaN  NaN     65.889055     2.101471   
height               201.0    NaN     NaN  NaN     53.766667     2.447822   
curb_weight          201.0    NaN     NaN  NaN   2555.666667   517.296727   
engine_type            201      6     ohc  145           NaN          NaN   
number_of_cylinders    201      7    four  157           NaN          NaN   
engine_size          201.0    NaN     NaN  NaN    126.875622    41.546834   
fuel_system            201      8    mpfi   92           NaN          NaN   
bore                 201.0    NaN     NaN  NaN      3.329701     0.268166   
stroke               201.0    NaN     NaN  NaN      3.261741     0.317875   
compression_ratio    201.0    NaN     NaN  NaN     10.164279     4.004965   
horsepower           201.0    NaN     NaN  NaN    103.263682    37.389372   
peak_rpm             201.0    NaN     NaN  NaN   5121.393035   479.624905   
city_mpg             201.0    NaN     NaN  NaN     25.179104      6.42322   
highway_mpg          201.0    NaN     NaN  NaN     30.686567      6.81515   
price                201.0    NaN     NaN  NaN  13207.129353  7947.066342   

                        min     25%      50%      75%      max  
symboling              -2.0     0.0      1.0      2.0      3.0  
normalized_losses      65.0   101.0    122.0    150.0    256.0  
make                    NaN     NaN      NaN      NaN      NaN  
fuel_type               NaN     NaN      NaN      NaN      NaN  
aspiration              NaN     NaN      NaN      NaN      NaN  
number_of_doors         NaN     NaN      NaN      NaN      NaN  
body_style              NaN     NaN      NaN      NaN      NaN  
drive_wheels            NaN     NaN      NaN      NaN      NaN  
engine_location         NaN     NaN      NaN      NaN      NaN  
wheel_base             86.6    94.5     97.0    102.4    120.9  
length                141.1   166.8    173.2    183.5    208.1  
width                  60.3    64.1     65.5     66.6     72.0  
height                 47.8    52.0     54.1     55.5     59.8  
curb_weight          1488.0  2169.0   2414.0   2926.0   4066.0  
engine_type             NaN     NaN      NaN      NaN      NaN  
number_of_cylinders     NaN     NaN      NaN      NaN      NaN  
engine_size            61.0    98.0    120.0    141.0    326.0  
fuel_system             NaN     NaN      NaN      NaN      NaN  
bore                   2.54    3.15     3.31     3.58     3.94  
stroke                 2.07    3.11     3.29     3.46     4.17  
compression_ratio       7.0     8.6      9.0      9.4     23.0  
horsepower             48.0    70.0     95.0    116.0    262.0  
peak_rpm             4150.0  4800.0   5200.0   5500.0   6600.0  
city_mpg               13.0    19.0     24.0     30.0     49.0  
highway_mpg            16.0    25.0     30.0     34.0     54.0  
price                5118.0  7775.0  10295.0  16500.0  45400.0

- The car price ranges from 5118 to 45400 units.
- The car weight ranges from 1488 to 4066 units.
- The most common car make in the data is of Toyota.

## Histogram

- A **histogram** is a univariate plot which helps us understand the distribution of a continuous numerical variable.
- It breaks the range of the continuous variables into a intervals of equal length and then counts the number of observations in each interval.
- We will use the *histplot()* function of seaborn to create histograms.

In [51]:
sns.histplot(data=df, x='price')
# plt.show()

<Axes: title={'center': 'Histogram:Price'}, xlabel='Price of cars', ylabel='Frequency'>

**Let's see how we can customize a histogram.**

In [49]:
plt.title('Histogram:Price')
plt.xlim(3000,50000)
plt.ylim(0,70)
plt.xlabel('Price of cars')
plt.ylabel('Frequency')
sns.histplot(data=df, x='price',color='orange');

**We can specify the number of intervals (or groups or bins) to create by setting the *bins* parameter.**

- If not specified it is passed to [numpy.histogram_bin_edges()](https://numpy.org/doc/stable/reference/generated/numpy.histogram_bin_edges.html#numpy.histogram_bin_edges)

In [ ]:
sns.histplot(data=df, x='price', bins=5)

In [ ]:
sns.histplot(data=df, x='price', bins=20)

**If we want to specify the width of the intervals (or groups or bins), we can use *binwidth* parameter.**

In [ ]:
sns.histplot(data=df, x='price', binwidth=20)

In [ ]:
sns.histplot(data=df, x='price', binwidth=200)

**How to find the optimal number of bins: Rule of thumb**

- We calculate the bin-width first, using the following formula: $$ binwidth =\frac{(2 * IQR)}{\sqrt[3]{n}} $$ where n = number of rows the dataset

- Then, we obtain bins using the calculated bin-width. $$ bins =\frac{Range}{binwidth} $$

**In addition to the bars, we can also add a density estimate by setting the *kde* parameter to *True*.**

- **Kernel Density Estimation**, or **KDE**, visualizes the distribution of data over a continuous interval.
- The conventional scale for KDE is: **Total frequency of each bin × Probability**

In [ ]:
sns.histplot(data=df, x='price', kde=True);

In [ ]:
sns.histplot(data=df, x='price', bins=700, kde=True);

Clearly, if we increase the number of bins, it reduces the frequency count in each group (bin). Since the scale of KDE depends on the total frequency of each bin (group), the above code gives us a flattened KDE plot.

**Let's check out the histograms for a few more attributes in the data.**

In [ ]:
sns.histplot(data=df, x='curb_weight', kde=True);

- A histogram is said to be **symmetric** if the left-hand and right-hand sides resemble mirror images of each other when the histogram is cut down the middle.

In [ ]:
sns.histplot(data=df, x='horsepower', kde=True);

- The tallest clusters of bars, i.e., peaks, in a histogram represent the **modes** of the data.
- A histogram **skewed to the right** has a large number of occurrences on the left side of the plot and a few on the right side of the plot.
- Similarly, a histogram **skewed to the left** has a large number of occurrences on the right side of the plot and few on the left side of the plot.

**Histograms are intuitive but it is hardly a good choice when we want to compare the distributions of several groups. For example,**

In [ ]:
sns.histplot(data=df, x='price', hue='body_style', kde=True);

It might be better to use subplots!

In [ ]:
g = sns.FacetGrid(df, col="body_style")
g.map(sns.histplot, "price");

In such cases, we can use **boxplots**. Boxplots, or box-and-whiskers plots, are an excellent way to visualize differences among groups.

## Boxplot

- A **boxplot**, or a **box-and-whisker plot**, shows the distribution of numerical data and skewness through displaying the data quartiles
- It is also called a **five-number summary plot**, where the five-number summary includes the minimum value, first quartile, median, third quartile, and the maximum value.
- The *boxplot()* function of seaborn can be used to create a boxplot.

In [ ]:
from IPython.display import Image
Image('/content/drive/MyDrive/Python Course/boxplot.png')
#Image('/location on your computer/boxplot.png')

In [ ]:
# creating a boxplot with seaborn
sns.boxplot(data=df, x='curb_weight');

**Let's see how we can customize a boxplot.**

In [ ]:
plt.title('Boxplot:Horsepower')
plt.xlim(30,300)
plt.xlabel('Horsepower')
sns.axes_style('whitegrid')
sns.boxplot(data=df, x='horsepower',color='green');

- In a boxplot, when the median is closer to the left of the box and the whisker is shorter on the left end of the box, we say that the distribution is **positively skewed (skewed right)**.
- Similarly, when the median is closer to the right of the box and the whisker is shorter on the right end of the box, we say that the distribution is **negatively skewed (skewed left)**.

In [ ]:
from IPython.display import Image
Image('/content/drive/MyDrive/skew_box.png')
#Image('/location on your computer/skew_box.png')

**For example,**

In [ ]:
sns.boxplot(data=df, x='price');

From the above plot, we can see that the distribution of `price` is positively skewed.

**Let's see how we can compare groups with boxplots.**

In [ ]:
sns.boxplot(data=df, x='body_style', y='price') ;

**Though boxplot visually summarizes variation in large datasets, it is unable to show multimodality and clusters.**

In [ ]:
sns.boxplot(data=df, x='bore');

- From the above boxplot we can not tell if the data is bimodal or not, but it is clearly visible in the following histogram.

In [ ]:
sns.histplot(data=df, x='bore',kde = True);

## Bar Graph

- A bar graph is generally used to show the counts of observations in each bin (or level or group) of categorical variable using bars.
- We can use the *countplot()* function of seaborn to plot a bar graph.

In [ ]:
sns.countplot(data=df, x='body_style');

**We can also make the plot more granular by specifying the *hue* parameter to display counts for subgroups.**

In [ ]:
sns.countplot(data=df, x='body_style', hue='fuel_type');

**Let's check out the bar graphs for a few more attributes in the data.**

In [ ]:
sns.countplot(data=df, x='make');

- This plot looks a little messy and congested.
- Let's increase the size of the plot to make it look better.

In [ ]:
plt.figure(figsize=(20,7))
sns.countplot(data=df, x='make');

- Some of the tick marks on the x-axis are overlapping with each other.
- Let's rotate the tick marks to make it look better.

In [ ]:
plt.figure(figsize=(20,7))
sns.countplot(data=df, x='make')
plt.xticks(rotation=90)

- A lot of plot-specific text has shown up in the output.
- Let's see how we can get rid of those.

In [ ]:
plt.figure(figsize=(20,7))
sns.countplot(data=df, x='make')
plt.xticks(rotation=90)
plt.show() # this will ensure that the plot is displayed without the text

**Here are some common ways to customize a barplot.**

In [ ]:
plt.figure(figsize=(10,7))
plt.title('Barplot:Engine-type')
plt.ylim(0,180)
sns.countplot(data=df, x='engine_type',hue='fuel_type')
plt.xlabel('Engine-type');

## Lineplot

Suppose, your dataset has multiple y values for each x value.  A lineplot is a great way to visualize this.  This type of data often shows up when we have data that evolves over time, for example, when we have monthly data over several years.  If we want to compare the individual months, then a line plot is a great option.  This is sometimes called seasonality analysis.



In [ ]:
from IPython.display import Image
Image('/content/drive/MyDrive/Python Course/Line_plot.png')
#Image('/location on your computer/Line_plot.png')

- A **line plot** uses straight lines to connect individual data points to display a trend or pattern in the data.
    - For example, seasonal effects and large changes over time.

- The *lineplot()* function of seaborn, by default, aggregates over multiple y values at each value of x and uses an estimate of the central tendency for the plot.

- *lineplot()* assumes that you are most often trying to draw y as a function of x. So, by default, it sorts the data by the x values before plotting.

In [ ]:
# loading one of the example datasets available in seaborn
flights = sns.load_dataset("flights")

# creating a line plot
sns.lineplot(data = flights , x = 'month' , y = 'passengers');

- The light blue shaded area is actually the '**confidence interval**' of the y-value estimates for each x-axis value.

- The **confidence interval** is a range of values around that estimate that are believed to contain the true value of that estimate with a certain probability.

**We can switch off the confidence intervals by setting the *ci* parameter to *'False'*.**

In [ ]:
sns.lineplot(data = flights , x = 'month' , y = 'passengers', errorbar=('ci', False));

**We can also check the relationship between two variables for different categories by specifying the *hue* parameter.**

In [ ]:
sns.lineplot(data=flights,x = 'month' , y = 'passengers', errorbar=('ci', False) ,hue='year');

**We can change the style of the lines by adding 'style' parameter to the function.**

In [ ]:
# loading one of the example datasets available in seaborn
fmri = sns.load_dataset("fmri")

# creating the line plot
sns.lineplot(data = fmri, x="timepoint", y="signal", hue="region", style="region", errorbar=('ci', False));

**We can also add markers at each observation to identify groups in a better way.**

In [ ]:
sns.lineplot(data = fmri, x="timepoint", y="signal", hue="region", style="region", errorbar=('ci', False), markers = True);

**Let's customize the lineplot for a better visualization.**

In [ ]:
plt.figure(figsize = (15,7))
sns.lineplot(data = flights , x = 'month' , y = 'passengers', hue = 'year')
plt.ylabel('Number of Passengers')
plt.legend(bbox_to_anchor=[1, 1]); #another way to change the legend's location in the plot

- Note that, unlike barplots and histograms, line plots may not include a zero baseline.
- We create a line chart is to emphasize changes in value, rather than the magnitude of the values themselves, and hence, a zero line is not meaningful.

## Scatterplot

Sometimes we want to know if two variables mean something when put together, whether a small change in one variable affects the other variable. In such cases, plotting a **scatterplot**, or **scatter-diagram**, with our data points can help us to check whether there is a potential relationship between them.

- A **scatterplot** is the simplest mode of a diagrammatic representation of two variables.
- It takes two perpendicular axes of coordinates, one for x and one for y.
- Unlike the lineplot, it directly plots each pair of values as a point on the 2D space.
- The *scatterplot()* function of seaborn can be used to make a scatterplot.

In [ ]:
sns.scatterplot(data=df, x='engine_size', y='horsepower');

**We can also check the relationship between two variables for different categories by specifying the *hue* parameter.**

In [ ]:
sns.scatterplot(data=df, x='engine_size', y='horsepower', hue='fuel_type');

**We can assign the same variable as *hue* to another parameter *style* which will vary the markers and create a more readable plot.**

In [ ]:
sns.scatterplot(data=df, x='engine_size', y='horsepower', hue='fuel_type', style='fuel_type');

**Correlation**

Correlation means association. More precisely, it expresses the extent to which two variables change together at a constant rate.

- In a scatter plot when the y variable tends to increase as the x variable increases, we say there is a **positive correlation** between the variables.
- Again, when the y variable tends to decrease as the x variable increases, we say there is a **negative correlation** between the variables.
- If the points on the scatter plot seem to be scattered randomly, we say that there is **no correlation** between the variables.

**Let's check out the relationship between a few more variables using scatter plots.**

In [ ]:
sns.scatterplot(data=df, x='curb_weight', y='engine_size');

From the above plot, we can say that these variables are *positively correlated.*

In [ ]:
sns.scatterplot(data=df, x='bore', y='stroke');

- From the above plot, it is clear that the variables have no correlation.

**Note:**
1. A strong correlation will have data points close together, while a weak correlation will have data points that are further apart.
2. We can not measure the relationship quantitatively using a scatter plot. It just gives an expression for the relative change between the variables.

We can see from the scatterplot of `engine_size` vs `horsepower` that there is a positive correlation between the two variables. Now, we want to measure the relationship between these two variables quantitatively and try to predict the '**horsepower**' based on '**engine size**'. This can be easily done by fitting a linear model. Here comes the seaborn ***lmplot()*** function to help us with that.

## lmplot

- The *lmplot()* function plots a scatterplot and also fit a linear model with a confidence interval for the fitted model.

In [ ]:
sns.lmplot(data=df, x='curb_weight', y='horsepower');

**We can also check the relationship between two variables for different categories by specifying the *hue* parameter.**

In [ ]:
sns.lmplot(data=df, x='curb_weight', y='horsepower', hue='fuel_type',ci=False);

**We can also disable the 95% confidence interval and just keep the regression model fit by setting the *ci* parameter to *False*.**

In [ ]:
sns.lmplot(data=df, x='curb_weight', y='horsepower', ci=False);

**We can also plot the levels of the third variable across different plots.**

In [ ]:
sns.lmplot(data=df, x='curb_weight', y='horsepower', col='fuel_type');

**Let's check out the relationship between a few more variables using *lmplot()*.**

In [ ]:
sns.lmplot(data=df, x='curb_weight', y='engine_size', col='number_of_doors', ci=False)

In [ ]:
sns.lmplot(data=df, x='horsepower', y='price', ci=False);

We have seen that histograms help us understand the distribution of individual variables and scatterplots help us identify the relationship between two variables. While we can view and analyze these plots separately, having them together in a single visualization would allow us to capture a lot more information in a concise manner. That's where a **jointplot** comes to our aid.

## Joint plot

* A jointplot helps to quickly visualize the bivariate and univariate profiles on the same plot.
* It comprises three plots - one displays the relationship between two variables while the other two show the individual distribution of each variable in the margins.
* The *jointplot()* function of seaborn can be used to create a jointplot.

In [ ]:
sns.jointplot(data=df, x='engine_size', y='horsepower');

- By default, *jointplot()* creates a scatter plot to show the relationship between two variables and individual histograms are created in x and y axis margins for each of the variables.

One of the drawbacks of **scatterplots** is the overlapping of points. When we have large volumes of data to plot, the data points in the plot overlap with each other, making it difficult to interpret the data.

In such cases, we can *divide the entire plot into bins using different shapes (square, triangle, hexagon, etc.)* and then try to see the number of data points falling within each of the bins.

seaborn's *jointplot()* provides a '**hex**' kind to plot the data in above-mentioned way.

**We can create a hexbin plot by setting *kind="hex"*.**

- It actually splits the plotting window into several hexagons and the colour of each hexagon denotes the number of data points in it.

In [ ]:
sns.jointplot(data=df, x='engine_size', y='horsepower', kind="hex");
plt.colorbar(); # adds a separate axis indicating the color scale in this plot

**We can also create a kde plot by setting *kind="kde"*.**.

- A KDE jointplot also uses color to determine where observations are the most dense.
- It creates a continuous plot using probabilities of new data should be introduced.

In [ ]:
sns.jointplot(data=df, x='engine_size', y='horsepower', kind="kde", fill=True);

**Let's check out the joint plots for a few more attributes in the data.**
- By using **kind = "reg"**, we can use the joint point to find the best line or curve that fits the plot.

In [ ]:
sns.jointplot(data=df, x='price', y='city_mpg', kind="reg");

## Violin Plot

* A **violin plot** allows us to visualize the distribution of a numeric variable for several groups.
* It is similar to **boxplot** but with a rotated plot on each side, giving more information about the density estimate on the y-axis.
* The *violinplot()* function of seaborn can be used to make a violin plot.

In [ ]:
sns.violinplot(data=df, x='horsepower');

- The white dot in the plot represents the **median**.
- The thick gray bar in the center represents the **interquartile range**.
- The wider section of the violin plot represents a higher probability and the skinnier sections represent a lower probability for the given value.

**We can get a vertical plot by setting *orient* parameter to *'v'* and assigning a numeric variable to the y-axis.**

In [ ]:
sns.violinplot(data=df, x='fuel_type', y='horsepower', orient='v');

**Let's check out the violinplots for a few more attributes in the data.**

In [ ]:
sns.violinplot(data=df, x='engine_size', y='fuel_type');

**We can use the *palette* parameter to change the colour palette.**

- seaborn has six variations of it - deep, muted, pastel, bright, dark, colorblind.

In [ ]:
sns.violinplot(data=df, hue='body_style', y='engine_size', palette="bright", legend = False);

In [ ]:
sns.violinplot(data=df, hue='engine_location', y='price', palette="colorblind", legend = False);

Histograms and other distribution plots are typically preferred for larger data sets. When we have a small dataset and we want to visualize the frequency distribution, a **strip plot** can be used.

## Strip Plot

* A strip plot is basically a scatter plot that differentiates different categories.
* The *stripplot()* function of seaborn can be used to make a strip plot.

In [ ]:
sns.stripplot(data=df, x='engine_size');

Strip plots are considered a good alternative to a box plot or a violin plot for comparing data distributions when we have fewer data points.

In [ ]:
plt.figure(figsize=(15,7))
sns.stripplot(data=df, x='body_style', y='engine_size');

**Strip plots are more useful when we add random noise called "jitter" to avoid overlapping of data points with same values.**

In [ ]:
plt.figure(figsize=(20,7))
sns.stripplot(data=df, x='body_style', y='engine_size', jitter=True);

**We can set the *hue* parameter to display observations for subgroups.**

In [ ]:
plt.figure(figsize=(10,7))
sns.stripplot(data=df, x='fuel_type', y='engine_size', hue="number_of_doors", jitter=True);

**Let's check out the strip plots for a few more attributes in the data.**

In [ ]:
plt.figure(figsize=(10,7))
sns.stripplot(data=df, x='number_of_doors', y='price',jitter=True);

In [ ]:
plt.figure(figsize=(10,7))
sns.stripplot(data=df, x='number_of_doors', y='horsepower', jitter=True);

In a strip plot, dots are randomly arranged in vertical direction, which means we can not use the width to estimate distribution. This can be easily done by **swarm plots**.

## Swarm Plot

* A **swarm plot** is very similar to the strip plot, but it avoids the overlapping of points.
* The *swarmplot( )* function of seaborn can be used to create such a graph for categorical values.

In [ ]:
sns.swarmplot(data=df, x='number_of_doors', y='price');

- In a swarm plot, dots are arranged in vertical columns.
- So, we can estimate the distribution of our data by looking at the width of the plot.

**We can add the *hue* parameter to *swarmplot()* and separate the categories.**

In [ ]:
sns.swarmplot(data=df, x='fuel_type', y='price', hue='number_of_doors');

**We can split the two products setting the *dodge* argument to *True*.**

In [ ]:
sns.swarmplot(data=df, x='fuel_type', y='price', hue='number_of_doors', dodge=True);

## Catplot

* The *catplot()* function provides a new framework giving access to several types of plots that show relationship between numerical variable and one or more categorical variables.
* Catplots supports 8 different plots including *stripplot()*, *swarmplot()*, *boxplot()* and more, with the default kind being 'strip'.

In [ ]:
sns.catplot(data=df, x='fuel_type', y='horsepower');

**We can set *kind='point'* to create a pointplot using the *catplot()* function.**

- A pointplot helps us to see how the main relationship is changing as a function of the second variable.

In [ ]:
sns.catplot(data=df, x='body_style', y='horsepower', hue='fuel_type', kind='point');

**Catplot is very useful to draw categorical plots onto a FacetGrid by assigning a third variable to the *col* parameter.**

In [ ]:
sns.catplot(data=df, x="fuel_type", y="horsepower", hue="number_of_doors", col="drive_wheels", kind='bar', palette='pastel');

**Let's check out the *catplot()* function for a few more attributes in the data.**

In [ ]:
sns.catplot(data=df, x="fuel_type", y="engine_size", hue="body_style", col="number_of_doors", kind='box', palette='bright');

In [ ]:
sns.catplot(data=df, x="number_of_doors", y="price", hue="body_style", col="fuel_type", kind='swarm', palette='muted');

## Pair Plot

* A **pairplot** shows the relationship between two numeric variables for each pair of columns in the dataset.
* It creates a grid of axes such that each variable in data will be shared in the y-axis across a single row and in the x-axis across a single column.
* The *pairplot()* function of seaborn can be used to create such a plot.

In [ ]:
sns.pairplot(data=df[['normalized_losses','wheel_base','curb_weight','engine_size','price','peak_rpm']])

**We can add the *hue* parameter in pairplot to create a semantic mapping.**

- It changes the default marginal plot to a layered kde plot.

**Also, we can add *vars* parameter to assign a list of variables from the dataset for which we want to create the pairplot.**

In [ ]:
sns.pairplot(data=df, vars=['wheel_base', 'curb_weight', 'engine_size', 'price'], hue='number_of_doors');

**We can set *corner=True* to plot only the lower triangle of a pairplot.**

In [ ]:
sns.pairplot(data=df, vars=['wheel_base', 'curb_weight', 'engine_size', 'price'], corner=True);

## Heatmap

* A **heatmap** is a graphical representation of data as a color-encoded matrix.
* It is a great way of representing  the correlation for each pair of columns in the data.
* The *heatmap()* function of seaborn helps us to create such a plot.

In [ ]:
sns.heatmap(data=df[['wheel_base','curb_weight','engine_size','price']].corr());

**We can set the *annot* parameter to *True* for displaying the numeric value in each cell.**

- To remove the color bar, the *cbar* parameter can be set to *False*.

In [ ]:
sns.heatmap(data=df[['wheel_base','curb_weight','engine_size','price']].corr(), annot=True, cbar=False);

**We can apply a different colormap with the *cmap* parameter for better visual appeal.**

In [ ]:
sns.heatmap(data=df[['wheel_base','curb_weight','engine_size','price']].corr(), annot=True, cmap='YlGnBu');

## Plotly

- **Plotly** is an open-source Python graphing library for building beautiful and interactive visualizations.

In [ ]:
# let's start by installing plotly
#!pip install plotly

In [ ]:
# importing plotly
import plotly.express as px

### Histogram

In [ ]:
his = px.histogram(df, x="price")
his.show()

### Bar Plot

In [ ]:
bar = px.bar(df, x='peak_rpm', y='horsepower')
bar.show()

### Scatter Plot

In [ ]:
scat = px.scatter(df, x='price', y='engine_size')
scat.show()

### Boxplot with underlying data

In [ ]:
fig = px.box(df, x="fuel_type", y="horsepower", points="all")
fig.show()

### 3D Scatter Plot

In [ ]:
fig_3d = px.scatter_3d(df, x='fuel_type', y='horsepower', z='price', color='horsepower')
fig_3d.show()

In [ ]:
# to save the output to an HTML file
fig_3d.write_html("scatter_3d.html")

***

## Customizing plots

- In this section we will see how we can customize  plots by using different matplotlib and seaborn functions.

To change the axis scales of a plot:

```
    plt.xlim(left, right)   # here left and right indicates the min and the max limits of  x-axis respectively
    plt.ylim(bottom, top)   # here bottom and top indicates the min and the max limits of  y-axis respectively
```

To change the plot title and axis labels:

```
    plt.title()
    plt.xlabel()   # for x_axis labels
    plt.ylabel()   # for y_axis labels
```

To show the grid lines of a plot:

```
    sns.set(style="darkgrid")
    sns.set(style="whitegrid")
    sns.set(style="white")
```

To hide the axes of a plot:

```
    plt.axis('off')   # it will hide both the x-axis and y-axis
```

To customize labels of the plot legend:

```
    ax.legend([])   # takes list of the legend values
```

To change the position of the plot legend:

```
    plt.legend(loc=___)
```

The following values can be added to *loc* to change the location of the legend:
- 'upper left', 'upper right', 'lower left', 'lower right', 'center', 'best', [etc.](https://matplotlib.org/stable/api/_as_gen/matplotlib.pyplot.legend.html)

***